### PR TITLE
[simple-app] Import all of the relevant PrometheusRules from the prometheus-alerts chart

### PR DESCRIPTION
**Pull request by diranged**
The goal of this PR is to make sure that each time a `simple-app` chart
is launched, it automatically includes relevant alerts for the
application itself. Many of these alerts are tricky because we have to
use `pod=~...` regex matches to scope things correctly.

The intention is that `simple-app` here will only have alerts relevant
to its own configuration... so if a chart uses several `simple-app`
charts, we'll end up with a different set of Prometheus Rules for each
chart.

### CODE REVIEW

File: charts/simple-app/templates/prometheusrules.yaml [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+          Pod {{`{{$labels.pod}}`}} in namespace {{`{{$labels.namespace}}`}}
+          has a container that has been terminated ({{`{{ $value }}`}} times) due to
+          {{`{{$labels.reason}}`}} in the last {{ .for }}.
+      expr: |-
+        sum by (container, namespace, pod, reason) (
+          sum_over_time(
+            kube_pod_container_status_terminated_reason{
+              reason=~"{{ join "|" .reasons }}",
+              namespace="{{ $targetNamespace }}",
+              pod=~"{{ include "simple-app.fullname" $global }}-.*"
```
Comment (sabw8217) [inline charts/simple-app/templates/prometheusrules.yaml:33]: if we are confident this pattern matching will always work, it seems a little sus to me

File: charts/simple-app/templates/prometheusrules.yaml [hunk 2]
```diff
@@ -0,0 +1,265 @@
...
+      annotations:
+        summary: Deployment generation mismatch due to possible roll-back
+        runbook_url: {{ $runbookUrl }}#deploymentgenerationmismatch
+        description: >-
+          Deployment generation for {{`{{`}} $labels.namespace
+          {{`}}`}}/{{`{{`}} $labels.deployment {{`}}`}} does not match, this
+          indicates that the Deployment has failed but has not been rolled
+          back.
+      expr: |-
+        kube_deployment_status_observed_generation{job="kube-state-metrics", namespace=~"{{ $targetNamespace }}"}
```
Comment (sabw8217) [inline charts/simple-app/templates/prometheusrules.yaml:171]: should have a filter on `deployment="{{ include "simple-app.fullname" . }}"`

File: charts/simple-app/templates/prometheusrules.yaml [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+          Deployment {{`{{`}} $labels.namespace {{`}}`}}/{{`{{`}}
+          $labels.deployment {{`}}`}} has not matched the expected number of
+          replicas for longer than {{ .for }}.
+      expr: |-
+        (
+          kube_deployment_spec_replicas{job="kube-state-metrics", namespace=~"{{ $targetNamespace }}"}
+            !=
+          kube_deployment_status_replicas_available{job="kube-state-metrics", namespace=~"{{ $targetNamespace }}"}
+        ) and (
+          changes(kube_deployment_status_replicas_updated{job="kube-state-metrics", namespace=~"{{ $targetNamespace }}"}[5m])
```
Comment (sabw8217) [inline charts/simple-app/templates/prometheusrules.yaml:197]: ditto

File: charts/simple-app/templates/prometheusrules.yaml [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        {{- with $values.prometheusRules.additionalRuleLabels }}
+        {{ toYaml . | nindent 8 }}
+        {{- end }}
+    {{- end }}
+
+  {{- if .Values.autoscaling.enabled }}
+  - name: {{ .Release.Namespace }}.{{ .Release.Name }}.{{ .Chart.Name }}.HorizontalPodAutoscalerRules
+    rules:
+
+    {{ with .Values.prometheusRules.HpaReplicasMismatch -}}
```
Comment (sabw8217) [inline charts/simple-app/templates/prometheusrules.yaml:213]: maybe should have `and .Values.autoscaling.enabled`?

File: charts/simple-app/templates/prometheusrules.yaml [hunk 5]
```diff
@@ -0,0 +1,265 @@
...
+        changes(kube_horizontalpodautoscaler_status_current_replicas[15m]) == 0
+      for: {{ .for }}
+      labels:
+        severity: {{ .severity }}
+        {{- with $values.prometheusRules.additionalRuleLabels }}
+        {{ toYaml . | nindent 8 }}
+        {{- end }}
+    {{- end }}
+
+    {{ with .Values.prometheusRules.HpaMaxedOut -}}
```
Comment (sabw8217) [inline charts/simple-app/templates/prometheusrules.yaml:244]: ditto

File: charts/simple-app/values.yaml [hunk 6]
```diff
@@ -482,3 +482,69 @@ runbookUrl:
...
+  # -- (`bool`) Whether or not to enable the container rules template
+  enabled: true
+
+  # -- (`map`) Additional custom labels attached to every PrometheusRule
+  additionalRuleLabels: {}
+
+  # -- Monitors Pods for Containers that are terminated either for unexpected
+  # reasons like ContainerCannotRun. If that number breaches the $threshold (1)
+  # for $for (1m), then it will alert.
+  PodContainerTerminated:
```
Comment (sabw8217) [inline charts/simple-app/values.yaml:501]: these should probably include an `enabled` flag to allow for granularly turning them off

File: charts/simple-app/templates/prometheusrules.yaml (hunk 2 above)
Comment (diranged) [reply charts/simple-app/templates/prometheusrules.yaml:171]: Good catch... fixing.

File: charts/simple-app/templates/prometheusrules.yaml (hunk 4 above)
Comment (diranged) [reply charts/simple-app/templates/prometheusrules.yaml:213]: @sabw8217 Thinking about this more ... I think this should be a global alarm. I can't think of a situation where a developer would have an autoscaling HPA in place, and want _different_ alarms than this?

File: charts/simple-app/Chart.yaml [hunk 7]
```diff
@@ -2,13 +2,15 @@ apiVersion: v2
...
 dependencies:
+
   - name: istio-alerts
-    version: 0.1.1
-    repository: "file://../istio-alerts/"
-    condition: istio-alerts.enabled
+    alias: istioRules
+    version: 0.1.2
+    repository: https://k8s-charts.nextdoor.com
+    condition: istioRules.enabled
```
Comment (diranged) [inline charts/simple-app/Chart.yaml:16]: All of the `istio` changes above are in #98... so I will rebase this change once we merge #98, which will then let the tests pass.

File: charts/simple-app/Chart.yaml [hunk 8]
```diff
@@ -2,13 +2,15 @@ apiVersion: v2
...
     email: matt@nextdoor.com
 dependencies:
+
   - name: istio-alerts
-    version: 0.1.1
-    repository: "file://../istio-alerts/"
-    condition: istio-alerts.enabled
+    alias: istioRules
+    version: 0.1.2
+    repository: https://k8s-charts.nextdoor.com
```
Comment (diranged) [inline charts/simple-app/Chart.yaml:15]: By using the user-facing repository endpoint, we no longer are forced to make a change  to all of the nested charts that depend on `istio-alerts` each time we update `istio-alerts`. Instead, each chart can be upgraded independently when we're ready. That seems better to me and less confusing.

File: charts/istio-alerts/templates/service-prometheusrule.yaml [hunk 9]
```diff
@@ -1,42 +1,50 @@
...
     nextdoor.com/source: {{ .Values.chart_source }}
 spec:
   groups:
-  - name: {{ .Release.Name }}.{{ .Release.Namespace }}.serviceRules
+  - name: {{ .Release.Namespace }}.{{ .Release.Name }}.{{ .Chart.Name }}.IstioServiceRules
     rules:
     {{ with .Values.serviceRules.http5XXMonitor }}
     {{ if .enabled }}
-    - alert: {{ $release.Name }}-5xx-Rate-Too-High
+    - alert: 5xx-Rate-Too-High
```
Comment (sabw8217) [inline charts/istio-alerts/templates/service-prometheusrule.yaml:18]: I like having the name of the service in the name of the alarm I get. `dev-changelog-5xx-rate-too-high` is a lot different than `us1-nextdoor-fe-fast-5xx-rate-too-high`!

File: charts/simple-app/templates/prometheusrules.yaml (hunk 1 above)
Comment (diranged) [reply charts/simple-app/templates/prometheusrules.yaml:33]: I think it works pretty well.. This is the full name of the deployment. So of course someone could find a way to create another deployment that matches the prefix, but at some point we can't really prevent that. Here's the scenario I could imagine:

  * App1: "myapp"
  * App2: "myapp-async"

In the above case, you could find that the alerts created for "myapp" apply to both "myapp-.*" and "myapp-async-.*" pods. I don't see a good way to get around that. However, as I write this, I can think of another scenario that is more likely to be hit:

Developer intentionally (or not) creates a nearly 63-char long deployment name by setting `.fullnameOverride`: "really-long-string-that-is-too-long-and-may-be-truncated-for-sure". When this happens, the replicaset automatically truncates part of the name:

```
$  k get deployments,rs,pods
NAME                                                                              READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/really-long-string                                                1/1     1            1           6m7s
deployment.apps/really-long-string-that-is-too-long-and-may-be-truncated-for-su   1/1     1            1           5m32s

NAME                                                                                         DESIRED   CURRENT   READY   AGE
replicaset.apps/really-long-string-7f8689d4f                                                 1         1         1       6m7s
replicaset.apps/really-long-string-that-is-too-long-and-may-be-truncated-for-su-5cf866cd98   1         1         1       5m32s

NAME                                                                  READY   STATUS    RESTARTS   AGE
pod/really-long-string-7f8689d4f-tlk4x                                1/1     Running   0          6m7s
pod/really-long-string-that-is-too-long-and-may-be-truncated-fn4z5j   1/1     Running   0          2m1s
```

In this case, we need to handle the fact that the pod-name is going to be truncated to attach a 5-character string at the end... so we need to truncate the prefix for the `pod=~...` selector to 58 characters.

File: charts/istio-alerts/templates/service-prometheusrule.yaml (hunk 9 above)
Comment (diranged) [reply charts/istio-alerts/templates/service-prometheusrule.yaml:18]: Sorry - we have two threads here... see https://github.com/Nextdoor/k8s-charts/pull/98#discussion_r783322249 for my response.

**Pull request merged**